README.md
MoveTools
=========

MoveTools is a plugin for [Deluge](http://deluge-torrent.org) that
shows the move status of a torrent and also gives the ability to force
a completed torrent to move to its assigned move completed path.

This is mainly a GtkUI plugin. It adds a status column and a torrent
submenu. WebUI support is minimal with only a status column added.


changelog
Version 0.2.0.2
- Add core initialization check in UI

Version 0.2.0.1
- Fix status rounding up to 100 percent

Version 0.2.0.0
- Progress estimation based on speed of previous transfers
- Sidebar filters for move status
- Cancel pending move operations

- Use reactor based updating rather than threads
- Explicitly process one move operation at a time

Version 0.1.8.1
- Deinit column last due to bug in Deluge column removal

Version 0.1.8
- Run progress calculations in separate thread
- Various fixes

Version 0.1.7
- Change column to use progress bar (adapted from Smart Move)
- Minor fixes

Version 0.1.6
- Detect attempts to move to the same path
- Minor fixes


movetools/core.py
#
# core.py
#
# Copyright (C) 2014 Ratanak Lun <ratanakvlun@gmail.com>
#
# Basic plugin template created by:
# Copyright (C) 2008 Martijn Voncken <mvoncken@gmail.com>
# Copyright (C) 2007-2009 Andrew Resch <andrewresch@gmail.com>
# Copyright (C) 2009 Damien Churchill <damoxc@gmail.com>
#
# Deluge is free software.
#
# You may redistribute it and/or modify it under the terms of the
# GNU General Public License, as published by the Free Software
# Foundation; either version 3 of the License, or (at your option)
# any later version.
#
# deluge is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with deluge.    If not, write to:
#   The Free Software Foundation, Inc.,
#   51 Franklin Street, Fifth Floor
#   Boston, MA  02110-1301, USA.
#
#    In addition, as a special exception, the copyright holders give
#    permission to link the code of portions of this program with the OpenSSL
#    library.
#    You must obey the GNU General Public License in all respects for all of
#    the code used other than OpenSSL. If you modify file(s) with this
#    exception, you may extend this exception to your version of the file(s),
#    but you are not obligated to do so. If you do not wish to do so, delete
#    this exception statement from your version. If you delete this exception
#    statement from all source files in the program, then also delete it here.
#


import time
import os
import os.path
import copy
import logging

from twisted.internet import reactor

from deluge.plugins.pluginbase import CorePluginBase
import deluge.component as component
import deluge.configmanager
from deluge.core.rpcserver import export
from deluge.core.torrent import Torrent

from common import PLUGIN_NAME
from common import MODULE_NAME
from common import STATUS_NAME
from common import STATUS_MESSAGE
from common import normalize_dict


CONFIG_FILE = "%s.conf" % MODULE_NAME

DEFAULT_PREFS = {
  "general": {
    "remove_empty": False,
    "estimated_speed": 20*2**20,
  },
  "timeout": {
    "success": -1.0,
    "error": -1.0,
  },
}

INIT_FILTERS = lambda: {
  "Moving": 0,
  "Queued": 0,
  "Done": 0,
  "Error": 0,
}

ALIVE_STATUS = ("Moving", "Queued")

UPDATE_INTERVAL = 2.0


log = logging.getLogger(__name__)


def get_total_size(paths):
  size = 0
  for path in paths:
    try:
      if os.path.exists(path):
        size += os.path.getsize(path)
    except OSError:
      pass

  return size


class Progress(object):

  def __init__(self, torrent, dest_path):
    self.torrent = torrent
    self._start_time = None
    self._end_time = None

    self.status = "Queued"
    self.message = "Queued"

    self.src_path = torrent.get_status(["save_path"])["save_path"]
    self.dest_path = dest_path

    files = torrent.get_files()

    src_paths = (os.path.join(self.src_path, f["path"]) for f in files)
    self.total_size = get_total_size(src_paths)

    self._paths = tuple(os.path.join(dest_path, f["path"]) for f in files)
    self.size = 0

    self.percent = 0.0
    self._estimated_speed = None

  def start(self, estimated_speed):
    self.status = "Moving"
    self.message = "Moving"
    self._start_time = time.time()
    self._estimated_speed = estimated_speed

  def finish(self):
    self._end_time = time.time()
    self.size = self.total_size
    self.percent = 100.0

  def get_elapsed(self):
    if self._end_time:
      elapsed = self._end_time - self._start_time
    else:
      elapsed = time.time() - self._start_time
    return elapsed

  def get_avg_speed(self):
    return self.size/(self.get_elapsed() or 1)

  def update(self):
    self._update_progress()
    self._update_status()

  def _update_progress(self):
    size = get_total_size(self._paths)
    if size == self.total_size:
      # OS reported full size, so use estimation
      size = self._estimated_speed * self.get_elapsed()
      if size > self.total_size:
        size = self.total_size

    self.size = size
    self.percent = float(self.size) / (self.total_size or 1) * 100

  def _update_status(self):
    if self.status == "Moving":
      if self.percent < 100.0:
        percent_str = ("%.6f" % self.percent)[:-4]
      else:
        percent_str = "99.99"

      self.message = "Moving %s" % percent_str


class Core(CorePluginBase):

  def enable(self):
    log.debug("[%s] Enabling Core...", PLUGIN_NAME)

    self.initialized = False

    if not deluge.component.get("TorrentManager").session_started:
      deluge.component.get("EventManager").register_event_handler(
        "SessionStartedEvent", self._on_session_started)
      log.debug("[%s] Waiting for session to start...", PLUGIN_NAME)
    else:
      reactor.callLater(0.1, self._initialize)

  def _on_session_started(self):
    log.debug("[%s] Resuming initialization...", PLUGIN_NAME)
    reactor.callLater(0.1, self._initialize)

  def _initialize(self):

    def move_storage(torrent, dest_path):
      id = str(torrent.handle.info_hash())
      log.debug("[%s] Queueing (%s)", PLUGIN_NAME, id)

      if id in self.torrents:
        if self.torrents[id].status in ALIVE_STATUS:
          log.debug("[%s] Unable to move torrent: already moving", PLUGIN_NAME)
          return False
        else:
          self._remove_job(id)

      self.torrents[id] = Progress(torrent, dest_path)

      if not dest_path:
        self._report_result(id, "error", "Error", "Empty path")
        return False

      if self.torrents[id].src_path == dest_path:
        self._report_result(id, "error", "Error", "Same path")
        return False

      self.queue.append(id)
      return True

    self.config = deluge.configmanager.ConfigManager(CONFIG_FILE,
      copy.deepcopy(DEFAULT_PREFS))

    self.general = self.config["general"]
    self.timeout = self.config["timeout"]

    normalize_dict(self.general, DEFAULT_PREFS["general"])
    normalize_dict(self.timeout, DEFAULT_PREFS["timeout"])

    self.torrents = {}
    self.calls = {}
    self.queue = []
    self.active = None

    component.get("AlertManager").register_handler("storage_moved_alert",
      self.on_storage_moved)
    component.get("AlertManager").register_handler(
      "storage_moved_failed_alert", self.on_storage_moved_failed)

    component.get("CorePluginManager").register_status_field(STATUS_NAME,
      self.get_move_status)
    component.get("CorePluginManager").register_status_field(STATUS_MESSAGE,
      self.get_move_message)

    component.get("FilterManager").register_tree_field(STATUS_NAME,
      INIT_FILTERS)

    self.orig_move_storage = Torrent.move_storage
    Torrent.move_storage = move_storage

    self.initialized = True

    log.debug("[%s] Core enabled", PLUGIN_NAME)

    self._update_loop()

  def disable(self):
    log.debug("[%s] Disabling Core...", PLUGIN_NAME)

    self.initialized = False

    deluge.component.get("EventManager").deregister_event_handler(
      "SessionStartedEvent", self._on_session_started)

    Torrent.move_storage = self.orig_move_storage

    for id in self.torrents:
      self._cancel_remove(id)

    component.get("FilterManager").deregister_tree_field(STATUS_NAME)

    component.get("CorePluginManager").deregister_status_field(STATUS_MESSAGE)
    component.get("CorePluginManager").deregister_status_field(STATUS_NAME)

    component.get("AlertManager").deregister_handler(
        self.on_storage_moved)
    component.get("AlertManager").deregister_handler(
        self.on_storage_moved_failed)

    self.config.save()
    deluge.configmanager.close(CONFIG_FILE)

    self._rpc_deregister(PLUGIN_NAME)

    log.debug("[%s] Core disabled", PLUGIN_NAME)

  @export
  def is_initialized(self):
    return self.initialized

  @export
  def set_settings(self, options):
    log.debug("[%s] Setting options", PLUGIN_NAME)
    self.general.update(options["general"])
    self.timeout.update(options["timeout"])

  @export
  def get_settings(self):
    log.debug("[%s] Getting options", PLUGIN_NAME)
    return {
      "general": self.general,
      "timeout": self.timeout,
    }

  @export
  def clear_selected(self, ids):
    log.debug("[%s] Clearing status results for: %s", PLUGIN_NAME, ids)
    for id in ids:
      if id in self.torrents and self.torrents[id].status not in ALIVE_STATUS:
        self._remove_job(id)

  @export
  def clear_all_status(self):
    log.debug("[%s] Clearing all status results", PLUGIN_NAME)
    for id in self.torrents.keys():
      if self.torrents[id].status not in ALIVE_STATUS:
        self._remove_job(id)

  @export
  def move_completed(self, ids):
    log.debug("[%s] Moving completed torrents in: %s", PLUGIN_NAME, ids)
    torrents = component.get("TorrentManager").torrents
    for id in ids:
      if id in torrents:
        torrent = torrents[id]
        if torrent.handle.is_finished():
          dest_path = torrent.options["move_completed_path"]
          torrent.move_storage(dest_path)

  @export
  def cancel_pending(self, ids):
    log.debug("[%s] Canceling pending move for: %s", PLUGIN_NAME, ids)
    for id in ids:
      if id in self.torrents and self.torrents[id].status == "Queued":
        self._remove_job(id)

  def on_storage_moved(self, alert):
    id = str(alert.handle.info_hash())
    if id in self.torrents:
      self.active = None
      self.torrents[id].finish()
      self._report_result(id, "success", "Done")

      if self.torrents[id].size >= self.config["general"]["estimated_speed"]*2:
        speed = self.torrents[id].get_avg_speed()
        self.config["general"]["estimated_speed"] = \
          int((self.config["general"]["estimated_speed"]*0.5 + speed*1.5)/2)
        log.debug("[%s] New estimated speed: %r B/s", PLUGIN_NAME,
          self.config["general"]["estimated_speed"])

      if self.general["remove_empty"]:
        try:
          log.debug("[%s] Removing empty folders in path: %s", PLUGIN_NAME,
            self.torrents[id].src_path)
          os.removedirs(self.torrents[id].src_path)
        except OSError:
          pass

  def on_storage_moved_failed(self, alert):
    id = str(alert.handle.info_hash())
    if id in self.torrents:
      self.active = None
      message = alert.message().rpartition(":")[2].strip()
      self._report_result(id, "error", "Error", message)

  def get_move_status(self, id):
    if id not in self.torrents:
      return None

    return self.torrents[id].status

  def get_move_message(self, id):
    if id not in self.torrents:
      return None

    return self.torrents[id].message

  def _update_loop(self):

    if not self.initialized:
      return

    if self.active is None:
      while self.queue:
        id = self.queue.pop(0)
        if id in self.torrents:
          job = self.torrents[id]
          if self.orig_move_storage(job.torrent, job.dest_path):
            log.debug("[%s] Moving (%s)", PLUGIN_NAME, id)
            job.start(self.config["general"]["estimated_speed"])
            self.active = id
            break

          self._report_result(id, "error", "Error", "General failure")

        self.active = None

    if self.active:
      self.torrents[self.active].update()

    reactor.callLater(UPDATE_INTERVAL, self._update_loop)

  def _report_result(self, id, type, status, message=""):
    if id in self.torrents:
      if message:
        message = "%s: %s" % (status, message)
      else:
        message = status

      log.debug("[%s] Status (%s): %s", PLUGIN_NAME, id, message)
      self.torrents[id].status = status
      self.torrents[id].message = message
      self._schedule_remove(id, self.timeout.get(type, 0))

  def _remove_job(self, id):
    self._cancel_remove(id)

    if id in self.queue:
      self.queue.remove(id)

    if id in self.torrents:
      del self.torrents[id]

  def _schedule_remove(self, id, time):
    self._cancel_remove(id)
    if time >= 0:
      self.calls[id] = reactor.callLater(time, self._remove_job, id)

  def _cancel_remove(self, id):
    if id in self.calls:
      if self.calls[id].active():
        self.calls[id].cancel()
      del self.calls[id]

  def _rpc_deregister(self, name):
    server = component.get("RPCServer")
    name = name.lower()

    for d in dir(self):
      if d[0] == "_": continue

      if getattr(getattr(self, d), '_rpcserver_export', False):
        method = "%s.%s" % (name, d)
        if method in server.factory.methods:
          log.debug("Deregistering method: %s", method)
          del server.factory.methods[method]


movetools/common.py
#
# common.py
#
# Copyright (C) 2014 Ratanak Lun <ratanakvlun@gmail.com>
#
# Basic plugin template created by:
# Copyright (C) 2008 Martijn Voncken <mvoncken@gmail.com>
# Copyright (C) 2007-2009 Andrew Resch <andrewresch@gmail.com>
# Copyright (C) 2009 Damien Churchill <damoxc@gmail.com>
#
# Deluge is free software.
#
# You may redistribute it and/or modify it under the terms of the
# GNU General Public License, as published by the Free Software
# Foundation; either version 3 of the License, or (at your option)
# any later version.
#
# deluge is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with deluge.    If not, write to:
#   The Free Software Foundation, Inc.,
#   51 Franklin Street, Fifth Floor
#   Boston, MA  02110-1301, USA.
#
#    In addition, as a special exception, the copyright holders give
#    permission to link the code of portions of this program with the OpenSSL
#    library.
#    You must obey the GNU General Public License in all respects for all of
#    the code used other than OpenSSL. If you modify file(s) with this
#    exception, you may extend this exception to your version of the file(s),
#    but you are not obligated to do so. If you do not wish to do so, delete
#    this exception statement from your version. If you delete this exception
#    statement from all source files in the program, then also delete it here.
#


import copy
import os
import pkg_resources


PLUGIN_NAME = "MoveTools"
MODULE_NAME = "movetools"
DISPLAY_NAME = _("MoveTools")

STATUS_NAME = _("Move Status")
STATUS_MESSAGE = "%s_message" % MODULE_NAME


def get_resource(filename):
  return pkg_resources.resource_filename(
      MODULE_NAME, os.path.join("data", filename))


def normalize_dict(dict_in, template):

  for key in dict_in.keys():
    if key not in template:
      del dict_in[key]

  for key in template:
    if key not in dict_in:
      dict_in[key] = copy.deepcopy(template[key])


def dict_equals(a, b):

  if len(a) != len(b):
    return False

  for key in a:
    if key not in b:
      return False

    if isinstance(a[key], dict):
      if not isinstance(b[key], dict):
        return False

      if a[key] is not b[key]:
        result = dict_equals(a[key], b[key])
        if not result:
          return False
    else:
      if a[key] != b[key]:
        return False

  return True


movetools/data/movetools.js
/*
Script: movetools.js
    The client-side javascript code for the MoveTools plugin.

Copyright:
    (C) Ratanak Lun 2014 <ratanakvlun@gmail.com>
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3, or (at your option)
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, write to:
        The Free Software Foundation, Inc.,
        51 Franklin Street, Fifth Floor
        Boston, MA  02110-1301, USA.

    In addition, as a special exception, the copyright holders give
    permission to link the code of portions of this program with the OpenSSL
    library.
    You must obey the GNU General Public License in all respects for all of
    the code used other than OpenSSL. If you modify file(s) with this
    exception, you may extend this exception to your version of the file(s),
    but you are not obligated to do so. If you do not wish to do so, delete
    this exception statement from your version. If you delete this exception
    statement from all source files in the program, then also delete it here.
*/


PLUGIN_NAME = 'MoveTools';
MODULE_NAME = 'movetools';
DISPLAY_NAME = _('MoveTools');

COLUMN_NAME = _('Move Status');

MoveToolsPlugin = Ext.extend(Deluge.Plugin, {

  name: PLUGIN_NAME,

  onDisable: function() {
    console.log('MoveToolsPlugin.onDisable');
    this.deregisterTorrentStatus(MODULE_NAME);
  },

  onEnable: function() {
    console.log('MoveToolsPlugin.onEnable');
    this.registerTorrentStatus(MODULE_NAME, COLUMN_NAME, {
      colCfg: {
        sortable: true
      }
    });
  }
});

Deluge.registerPlugin(PLUGIN_NAME, MoveToolsPlugin);


movetools/gtkui.py
#
# gtkui.py
#
# Copyright (C) 2014 Ratanak Lun <ratanakvlun@gmail.com>
#
# Basic plugin template created by:
# Copyright (C) 2008 Martijn Voncken <mvoncken@gmail.com>
# Copyright (C) 2007-2009 Andrew Resch <andrewresch@gmail.com>
# Copyright (C) 2009 Damien Churchill <damoxc@gmail.com>
#
# Deluge is free software.
#
# You may redistribute it and/or modify it under the terms of the
# GNU General Public License, as published by the Free Software
# Foundation; either version 3 of the License, or (at your option)
# any later version.
#
# deluge is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with deluge.    If not, write to:
#   The Free Software Foundation, Inc.,
#   51 Franklin Street, Fifth Floor
#   Boston, MA  02110-1301, USA.
#
#    In addition, as a special exception, the copyright holders give
#    permission to link the code of portions of this program with the OpenSSL
#    library.
#    You must obey the GNU General Public License in all respects for all of
#    the code used other than OpenSSL. If you modify file(s) with this
#    exception, you may extend this exception to your version of the file(s),
#    but you are not obligated to do so. If you do not wish to do so, delete
#    this exception statement from your version. If you delete this exception
#    statement from all source files in the program, then also delete it here.
#


import logging

import gtk
import gtk.glade

from twisted.internet import reactor

from deluge.ui.client import client
from deluge.plugins.pluginbase import GtkPluginBase
import deluge.component as component

from common import PLUGIN_NAME
from common import MODULE_NAME
from common import DISPLAY_NAME
from common import STATUS_NAME
from common import STATUS_MESSAGE
from common import get_resource
from common import dict_equals

INIT_POLLING_INTERVAL = 3.0


log = logging.getLogger(__name__)


class GtkUI(GtkPluginBase):

  def enable(self):
    log.debug("[%s] Enabling GtkUI...", PLUGIN_NAME)
    self._poll_init()

  def _poll_init(self):
    client.movetools.is_initialized().addCallback(self._check_init)

  def _check_init(self, result):
    log.debug("[%s] Waiting for core to be initialized...", PLUGIN_NAME)
    if result == True:
      self._finish_init()
    else:
      reactor.callLater(INIT_POLLING_INTERVAL, self._poll_init)

  def _finish_init(self):
    log.debug("[%s] Resuming initialization...", PLUGIN_NAME)

    self.config = {}

    self.ui = gtk.glade.XML(get_resource("wnd_preferences.glade"))

    lbl = self.ui.get_widget("lbl_general")
    lbl.set_markup("<b>%s</b>" % lbl.get_text())

    lbl = self.ui.get_widget("lbl_timeout")
    lbl.set_markup("<b>%s</b>" % lbl.get_text())

    component.get("Preferences").add_page(
        DISPLAY_NAME, self.ui.get_widget("blk_preferences"))
    component.get("PluginManager").register_hook(
        "on_apply_prefs", self._do_save_settings)
    component.get("PluginManager").register_hook(
        "on_show_prefs", self._do_load_settings)

    self.menu = self._create_menu()
    self.menu.show_all()

    self.sep = component.get("MenuBar").add_torrentmenu_separator()
    component.get("MenuBar").torrentmenu.append(self.menu)

    self._add_column()

    self._do_load_settings()
    log.debug("[%s] GtkUI enabled", PLUGIN_NAME)

  def disable(self):
    log.debug("[%s] Disabling GtkUI...", PLUGIN_NAME)

    component.get("MenuBar").torrentmenu.remove(self.sep)
    component.get("MenuBar").torrentmenu.remove(self.menu)

    self.menu.destroy()

    component.get("Preferences").remove_page(DISPLAY_NAME)
    component.get("PluginManager").deregister_hook(
        "on_apply_prefs", self._do_save_settings)
    component.get("PluginManager").deregister_hook(
        "on_show_prefs", self._do_load_settings)

    self._remove_column()

    log.debug("[%s] GtkUI disabled", PLUGIN_NAME)

  def _do_save_settings(self):
    log.debug("[%s] Saving settings", PLUGIN_NAME)

    config = {
      "general": {
        "estimated_speed":
          self.ui.get_widget("spn_estimated_speed").get_value(),
        "remove_empty": self.ui.get_widget("chk_remove_empty").get_active(),
      },
      "timeout": {
        "success": self.ui.get_widget("spn_success_timeout").get_value(),
        "error": self.ui.get_widget("spn_error_timeout").get_value(),
      },
    }

    if not dict_equals(config, self.config):
      client.movetools.set_settings(config)
    else:
      log.debug("[%s] No settings were changed", PLUGIN_NAME)

  def _do_load_settings(self):
    log.debug("[%s] Requesting settings", PLUGIN_NAME)
    client.movetools.get_settings().addCallback(self._do_load)

  def _do_load(self, config):
    self.config = config

    spn = self.ui.get_widget("spn_estimated_speed")
    spn.set_value(config["general"]["estimated_speed"])
    chk = self.ui.get_widget("chk_remove_empty")
    chk.set_active(config["general"]["remove_empty"])

    spn = self.ui.get_widget("spn_success_timeout")
    spn.set_value(config["timeout"]["success"])
    spn = self.ui.get_widget("spn_error_timeout")
    spn.set_value(config["timeout"]["error"])

  def _create_menu(self):
    menu = gtk.MenuItem(DISPLAY_NAME)
    submenu = gtk.Menu()

    move_item = gtk.MenuItem(_("Move"))
    move_submenu = gtk.Menu()
    move_item.set_submenu(move_submenu)

    item = gtk.MenuItem(_("Move Completed"))
    item.connect("activate", self._do_move_completed)
    move_submenu.append(item)

    item = gtk.MenuItem(_("Cancel Pending"))
    item.connect("activate", self._do_cancel_pending)
    move_submenu.append(item)

    status_item = gtk.MenuItem(_("Status"))
    status_submenu = gtk.Menu()
    status_item.set_submenu(status_submenu)

    item = gtk.MenuItem(_("Clear"))
    item.connect("activate", self._do_clear_selected)
    status_submenu.append(item)

    item = gtk.MenuItem(_("Clear All"))
    item.connect("activate", self._do_clear_all)
    status_submenu.append(item)

    submenu.append(move_item)
    submenu.append(status_item)
    menu.set_submenu(submenu)

    return menu

  def _do_move_completed(self, widget):
    ids = component.get("TorrentView").get_selected_torrents()
    log.debug("[%s] Requesting move completed for: %s", PLUGIN_NAME, ids)
    client.movetools.move_completed(ids)

  def _do_cancel_pending(self, widget):
    ids = component.get("TorrentView").get_selected_torrents()
    log.debug("[%s] Requesting cancel pending for: %s", PLUGIN_NAME, ids)
    client.movetools.cancel_pending(ids)

  def _do_clear_selected(self, widget):
    ids = component.get("TorrentView").get_selected_torrents()
    log.debug("[%s] Requesting clear status results for: %s",
        PLUGIN_NAME, ids)
    client.movetools.clear_selected(ids)

  def _do_clear_all(self, widget):
    log.debug("[%s] Requesting clear all status results", PLUGIN_NAME)
    client.movetools.clear_all_status()

  def _add_column(self):
    renderer = gtk.CellRendererProgress()

    component.get("TorrentView").add_column(
      header=STATUS_NAME,
      render=renderer,
      col_types=str,
      hidden=False,
      position=None,
      status_field=[STATUS_MESSAGE],
      function=self._render_cell,
      sortid=0,
      column_type="progress",
    )

  def _render_cell(self, column, cell, model, iter, data):
    cell.set_property("value", 0.0)
    cell.set_property("visible", False)

    status = model[iter][data[0]]
    if status:
      cell.set_property("visible", True)

      if status.startswith("Moving "):
        try:
          value = float(status.split()[-1])
          cell.set_property("value", value)

          status = "%s %.2f%%" % (_("Moving"), value)
        except ValueError:
          status = _("Status error")

        cell.set_property("text", status)
      else:
        if status == "Done":
          cell.set_property("value", 100.0)

        cell.set_property("text", _(status))

  def _remove_column(self):
    component.get("TorrentView").remove_column(STATUS_NAME)
